examples/instance/maas/instance.go

```diff
@@ -0,0 +1,282 @@
+
+	if spec.Properties != nil {
+		if err := spec.Properties.Decode(&properties); err != nil {
+			return nil, fmt.Errorf("Invalid instance properties: %s", err)
+		}
+	}
+	nodeListing := m.MaasObj.GetSubObject("nodes")
+	jsonResponse, err := nodeListing.CallPost("acquire", m.convertSpecToMaasParam(properties))
+	if err != nil {
+		return nil, err
+	}
+	acquiredNode, err := jsonResponse.GetMAASObject()
+	if err != nil {
+		return nil, err
+	}
+	listNodeObjects, err := nodeListing.CallGet("list", url.Values{})
+	if err != nil {
+		return nil, err
+	}
+	listNodes, err := listNodeObjects.GetArray()
+	if err != nil {
+		return nil, err
+	}
+	notDuplicate := false
+	for range listNodes {
+		hostname, err := acquiredNode.GetField("hostname")
+		if err != nil {
+			return nil, err
+		}
+		isdup, err := m.checkDuplicate(hostname)
+		if err != nil {
+			return nil, err
+		}
+		if isdup {
+			jsonResponse, err = nodeListing.CallPost("acquire", m.convertSpecToMaasParam(properties))
+			if err != nil {
+				return nil, err
+			}
+			acquiredNode, err = jsonResponse.GetMAASObject()
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			notDuplicate = true
+			break
+		}
+	}
+	if !notDuplicate {
+		return nil, errors.New("Failed to aquire node")
+	}
+	isAcquired := make(chan bool)
+	go func() {
+		for state, err := acquiredNode.GetField("substatus_name"); state != "Allocated" && err == nil; state, err = acquiredNode.GetField("substatus_name") {
+			time.Sleep(100)
+		}
+		isAcquired <- true
+		return
+	}()
+	<-isAcquired
+	params := url.Values{}
+	if _, err = acquiredNode.CallPost("start", params); err != nil {
+		return nil, err
+	}
+	hostname, err := acquiredNode.GetField("hostname")
+	if err != nil {
+		return nil, err
+	}
+	id := instance.ID(hostname)
+	machineDir, err := ioutil.TempDir(m.MaasfilesDir, "infrakit-")
+	if err != nil {
+		return nil, err
+	}
+	if err := ioutil.WriteFile(path.Join(machineDir, "MachineName"), []byte(hostname), 0755); err != nil {
+		return nil, err
+	}
+	tagData, err := types.AnyValue(spec.Tags)
+	if err != nil {
+		return nil, err
+	}
+	if err := ioutil.WriteFile(path.Join(machineDir, "tags"), tagData.Bytes(), 0666); err != nil {
+		return nil, err
+	}
+	if spec.LogicalID != nil {
+		if err := ioutil.WriteFile(path.Join(machineDir, "ip"), []byte(*spec.LogicalID), 0666); err != nil {
+			return nil, err
+		}
+	}
+	return &id, nil
+}
+
+// Label labels the instance
+func (m maasPlugin) Label(instance instance.ID, labels map[string]string) error {
+	machineDir := path.Join(m.MaasfilesDir, string(instance))
+	tagFile := path.Join(machineDir, "tags")
+	buff, err := ioutil.ReadFile(tagFile)
+	if err != nil {
+		return err
+	}
+
+	tags := map[string]string{}
+	err = types.AnyBytes(buff).Decode(&tags)
+	if err != nil {
+		return err
+	}
+
+	for k, v := range labels {
+		tags[k] = v
+	}
+
+	encoded, err := types.AnyValue(tags)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(tagFile, encoded.Bytes(), 0666)
```

Comment:
Files are written locally to store tags...  is there a way in MAAS to tag resources and query by tags?  https://docs.ubuntu.com/maas/2.1/en/intro-concepts#tags  

If we can reduce amount of state we need store locally, the better.     If there are ways to do tagging via the MAAS api, then we should consider making that change in a future PR?